src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            head.prev = new Node<>(null, value, head);
+            head = head.prev;
+            size++;
+        } else {
+            Node<T> indexNode = getNode(index);
+            Node<T> current = new Node<>(indexNode.prev, value, indexNode);
+            current.next.prev = current;
+            current.prev.next = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T replacedValue = getNode(index).value;
+        getNode(index).value = value;
+        return replacedValue;
```

Comment:
you are searching twice for a node with the same index
```suggestion
        Node<T> node = getNode(index);
        T replacedValue = node.value;
        node.value = value;
        return replacedValue;
```